pkgs/build-support/rust/build-rust-crate/default.nix

```diff
@@ -46,10 +46,21 @@ let
             else
               extern;
           opts = lib.optionalString (dep.stdlib or false) "noprelude:";
+
+          # To ensure cross compilation works as expected (i.e for
+          # macOS -> Linux cross compilation where the shared library
+          # extension is .dylib for macOS and .so for Linux),
+          # we should select the appropriate shared library extension.
+          # Proc macros are compiled/run on the build platform architecture.
+          sharedLibraryExtension =
+            if lib.attrByPath ["procMacro"] false dep
+            then stdenv.buildPlatform.extensions.sharedLibrary
+            else stdenv.hostPlatform.extensions.sharedLibrary;
+
           filename =
             if lib.any (x: x == "lib" || x == "rlib") dep.crateType
             then "${dep.metadata}.rlib"
-            else "${dep.metadata}${stdenv.hostPlatform.extensions.sharedLibrary}";
+            else "${dep.metadata}${sharedLibraryExtension}";
```

Comment:
In https://github.com/NixOS/nixpkgs/pull/319153 I changed it to use `extensions.library`, to avoid conflicts can you use that instead and rename the let binding? Also, it seems there are other references to sharedLibrary/library in other places in buildRustCrate, do you not need to change them too? I managed to cross compile Linux to wasm with proc macros no problem, is this only with Darwin <> Linux?

Reply:
Thanks for mentioning that. I'll gave a look when I get a chance.

This was an issue we only daw of macos -> linux, yes. We don't do wasm compilation so I can't say.

Reply:
@flokli and I just encountered the same issue when building Darwin -> wasm32